WIP: plausible deniability

Fabian-K opened the pull request:
🏗

kolyaopahle committed:
Added requestPadding property to all VerficationService requests

kolyaopahle committed:
Added fake calls to the WebReqeustBuilder

Fabian-K committed:
🚧 initial draft to enable plausible deniability

kolyaopahle committed:
Merge branch 'feature/fake-request-support' into feature/plausible-deniability

kolyaopahle committed:
Switched from SubmissionConstants to KeyType enum everywhere

kolyaopahle committed:
basic playbook implementation with fake and real requests

Fabian-K committed:
Playbook
- ensure request pattern for playbooks is always the same

VerificationService
- apply padding to ensure equal request size (header & body)

SecurityHelper
- extract hash256 to HashHelper. This simplifies tests that use only the hash function (and therefore don´t need to initialize SecurityHelper and its dependencies)

Fabian-K committed:
Playbook
- ignore exceptions for fake requests

SubmissionService
- add padding header to fake request for same header size

WebRequestBuilder
- include fake keys in upload (🏗)

kolyaopahle committed:
Implemented random chance of dummy playbook execution on app open

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

Fabian-K committed:
DiagnosisKeyService: removed (low value & difficult to test)

SubmissionService & SubmitDiagnosisKeysTransaction
- inline playbook & backgroundNoise property to prevent issues during testing

DiagnosisKeyConstantsTest, SubmissionServiceTest, SubmitDiagnosisKeysTransactionTest & SubmissionViewModelTest
- adjusted to changes

kolyaopahle committed:
Dummy playbook will now be repeated and delayed randomly

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge branch 'feature/plausible-deniability' of https://github.com/corona-warn-app/cwa-app-android into feature/plausible-deniability

 Conflicts:
	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt

kolyaopahle committed:
Linting

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Initial Code for background noise worker

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
First implementation of noise background worker

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Linting

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

Fabian-K committed:
Merge remote-tracking branch 'origin/dev' into feature/plausible-deniability

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt

Fabian-K committed:
PlaybookImpl
- ensure that fake requests are executed when real requests fail

SubmissionService & VerificationService
- adjust header name for padding

WebRequestBuilder
- add padding to dummy submission

Fabian-K committed:
BackgroundNoise is now trigger-only

PlaybookImpl
- include follow-up executions after every playbook
- logging

SubmissionViewModel.kt, SubmissionService.kt, SubmitDiagnosisKeysTransaction.kt, MainActivity.kt, BackgroundNoisePeriodicWorker.kt, DiagnosisTestResultRetrievalPeriodicWorker.kt
- propagate context for coroutine

VerificationService
- ensure body size of 1000

Fabian-K committed:
WebRequestBuilder.kt
- adjust fake key generation

PlaybookImplTest.kt
- remove unused server.enqueue

SubmissionService.kt 6 SubmitDiagnosisKeysTransaction.kt
- remove commented out code

Fabian-K committed:
Merge branch 'dev' into feature/plausible-deniability

Fabian-K committed:
revert temporary changes to SubmissionResultPositiveOtherWarningFragment.kt

Fabian-K committed:
- adjust fake key size
- remove temporary comment

kolyaopahle committed:
Background job scheduling implemented

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge remote-tracking branch 'origin/feature/plausible-deniability' into feature/plausible-deniability

kolyaopahle committed:
Moved build work calls to own file to fix linting

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge remote-tracking branch 'origin/feature/plausible-deniability' into feature/plausible-deniability

jakobmoellersap commented on Corona-Warn-App/build.gradle:
```diff
@@ -244,11 +244,13 @@ dependencies {
     // TESTING
     testImplementation 'junit:junit:4.13'
     testImplementation "android.arch.core:core-testing:1.1.1"
+    testImplementation 'org.jetbrains.kotlinx:kotlinx-coroutines-test:1.3.7'
```
this is still experimental and could cause issues. TBD

jakobmoellersap commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/http/playbook/PlaybookImpl.kt:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    private suspend fun <T> executeCapturingExceptions(body: suspend () -> T): Pair<T?, Exception?> {
```
Maybe a simple throw can be enough here depending on the use case.

Fabian-K replied:
We need to ensure that following fake requests of the playbook are still executed even when the real request fails

jakobmoellersap commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/http/service/VerificationService.kt:
```diff
@@ -17,20 +17,23 @@ interface VerificationService {
     suspend fun getTAN(
         @Url url: String,
         @Header("cwa-fake") fake: String,
+        @Header("cwa-header-padding") headerPadding: String?,
```
Is there a case where we do not have padding?

Fabian-K replied:
I think currently you are right. The header can be omitted when the header of a request happens to be exactly as big as we need it.

Fabian-K committed:
- initialize coroutine scope within the playbook, revert passing it from outside
- remove experimental test dependency for coroutines

Fabian-K committed:
Merge branch 'dev' into feature/plausible-deniability

Fabian-K commented:
Thank you for the review. Feedback is applied with 863b12d

Fabian-K committed:
- use single endpoint per server for fake requests
- reduce request size from 1000 to 250 for the verification server
- include dummy registration token in fake request to fulfill verification on server side
- prepare for randomized count of submitted keys
- always include headers cwa-authorization & cwa-header-padding for submission server

Fabian-K committed:
- simplify empty header using constant

Fabian-K committed:
Merge remote-tracking branch 'origin/dev' into feature/plausible-deniability

Fabian42 commented:
I'm interested: What does this change do? An emoji is not a very good description…

vaubaehn commented:
@Fabian42 : as far as I can interpret the code, a new function for data privacy is introduced. Randomly when opening the app, there is a fake TEK uploading to the server. It is to protect a COVID positive tested user, who uploads his keys to the server to warn the others, while his network traffic is under oberservation (e. g., by a hacker...) By this, he may deny (upon being confronted with 'his network behaviour') that he uploaded keys, because ALL CWA apps now uploading (fake) keys from time. So, network sniffing is not reliable anymore to identify positive tested users.

Dear devs, please correct me if I'm wrong.

vaubaehn commented:
@Fabian42 ... sorry, now realized that you most likely don't need any explanation about the code... Your original question had a different intention, I guess... ;)

Fabian42 commented:
No, that is actually exactly what I asked about. Thanks!
